### PR TITLE
Issue/7696 fix agent cache

### PR DESCRIPTION
**Pull request by Hugo-Inmanta**
# Description

part 1 of https://github.com/inmanta/inmanta-core/issues/7696

- externalize cleanup: cache is now periodically cleaned up by the `cleanup_stale_cache_entries` job.
- convert `timerqueue` to a heap queue

part 2 [PR](https://github.com/inmanta/inmanta-core/pull/7850)
- converts version-based caching to time based

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
~~- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )~~
~~- [ ] If this PR fixes a race condition in the test suite, also push the fix to the relevant stable branche(s) (see [test-fixes](https://internal.inmanta.com/development/core/tasks/build-master.html#test-fixes) for more info)~~

### CODE REVIEW

File: src/inmanta/agent/forking_executor.py [hunk 1]
```diff
@@ -287,8 +287,21 @@ async def call(self, context: ExecutorContext) -> typing.Sequence[inmanta.loader
                 logger.info("Failed to load sources: %s", module_source, exc_info=True)
                 failed.append(module_source)
 
+        self.periodic_cache_cleanup_job = asyncio.create_task(self.cleanup_stale_cache_entries(context))
+
         return failed
 
+    async def cleanup_stale_cache_entries(self, context: ExecutorContext) -> None:
```
Comment (wouterdb) [inline src/inmanta/agent/forking_executor.py:294]: I would expect this code to be in the local executor, as it is the one actually contains it

**Review by wouterdb — CHANGES_REQUESTED** [review]

File: src/inmanta/agent/cache.py [hunk 2]
```diff
@@ -40,12 +40,21 @@ def __init__(self, timeout: int = 24 * 3600, version: int = 0) -> None:
...
+        value: Any,
+        call_on_delete: Optional[Callable[[Any], None]],
+        freshness_period: float,
+    ) -> None:
         self.key = key
         self.scope = scope
         self.value = value
         self.time: float = time.time() + scope.timeout
         self.call_on_delete = call_on_delete
+        self.last_used_at = datetime.datetime.now().astimezone()
```
[the remark this answers is not in the document]
Comment (Hugo-Inmanta) [reply src/inmanta/agent/cache.py]: Changed this to:
- the cache item will mark itself as stale after the given `freshness_period`
- this will put this key in a list of keys to cleanup

During cleanup, under exclusive work_in_progress lock:
- go through this list and remove this item from the cache (if they're still in it)

During work (deploy/dry_run/get_facts), under exclusive work_in_progress lock:
- if we hit a stale item, synchronously remove it from the cache before recreating it

File: src/inmanta/agent/cache.py [hunk 3]
```diff
@@ -40,13 +40,24 @@ def __init__(self, timeout: int = 24 * 3600, version: int = 0) -> None:
 
 
 class CacheItem:
-    def __init__(self, key: str, scope: Scope, value: Any, call_on_delete: Optional[Callable[[Any], None]]) -> None:
```
Comment (wouterdb) [inline src/inmanta/agent/cache.py:43]: This while cache evalidation is getting too complicated, I would revert this to what it was and rename `_advance_time` to `clean_stale_entries`  and remove all calls to `_advance_time`

File: src/inmanta/agent/in_process_executor.py [hunk 4]
```diff
@@ -67,12 +67,29 @@ def __init__(
...
+        self.periodic_cache_cleanup_job: Optional[asyncio.Task[None]] = None
+
+    async def start(self) -> None:
+        self.periodic_cache_cleanup_job = asyncio.create_task(self.cleanup_stale_cache_entries())
+
+    async def cleanup_stale_cache_entries(self) -> None:
+        """
+        This task periodically cleans up stale entries in the cache
+        """
+        while True:
```
Comment (wouterdb) [inline src/inmanta/agent/in_process_executor.py:88]: when stopping, this should stopped as well (and awaited in join)

File: src/inmanta/agent/in_process_executor.py [hunk 5]
```diff
@@ -67,12 +67,29 @@ def __init__(
...
+            reschedule_interval: float = 1.0
+            async with self.wip_lock:
+                self._cache.clean_stale_entries()
+            await asyncio.sleep(reschedule_interval)
```
Comment (wouterdb) [inline src/inmanta/agent/in_process_executor.py:92]: A few lines below, the `self._cache.close()` should probably go on the threadpool as well  (requiring stop to be an async method), because it is called from the ioloop and it calls the finalizers

File: tests/conftest.py [hunk 6]
```diff
@@ -1937,3 +1938,22 @@ def disable_version_and_agent_cleanup_job():
     orchestrationservice.PERFORM_CLEANUP = False
     yield
     orchestrationservice.PERFORM_CLEANUP = old_perform_cleanup
+
+
+@pytest.fixture(scope="session")
+def pip_index(tmp_path_factory) -> utils.PipIndex:
```
Comment (wouterdb) [inline tests/conftest.py:1944]: Why is this added?

File: src/inmanta/agent/in_process_executor.py [hunk 7]
```diff
@@ -67,12 +67,29 @@ def __init__(
...
+        self.periodic_cache_cleanup_job = asyncio.create_task(self.cleanup_stale_cache_entries())
+
+    async def cleanup_stale_cache_entries(self) -> None:
+        """
+        This task periodically cleans up stale entries in the cache
+        """
+        while True:
+            reschedule_interval: float = 1.0
+            async with self.wip_lock:
+                self._cache.clean_stale_entries()
```
Comment (wouterdb) [inline src/inmanta/agent/in_process_executor.py:91]: This has to go on the threadpool

**Review by wouterdb — APPROVED** [review]

File: tests/conftest.py (hunk 6 above)
Comment (Hugo-Inmanta) [reply tests/conftest.py]: Not needed -> removed it

File: src/inmanta/agent/cache.py [hunk 8]
```diff
@@ -202,11 +221,7 @@ def _cache(self, item: CacheItem) -> None:
                 self.keysforVersion[scope.version].add(item.key)
             except KeyError:
                 raise Exception("Added data to version that is not open")
-
-        bisect.insort_right(self.timerqueue, item)
-        if item.time < self.nextAction:
-            self.nextAction = item.time
```
Comment (wouterdb) [inline src/inmanta/agent/cache.py:208]: this is still needed

**Review by wouterdb — APPROVED** [review]

File: src/inmanta/agent/cache.py (hunk 8 above)
Comment (Hugo-Inmanta) [reply src/inmanta/agent/cache.py:208]: indeed :+1:

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: src/inmanta/agent/in_process_executor.py [hunk 9]
```diff
@@ -68,20 +68,42 @@ def __init__(
         self.thread_pool: ThreadPoolExecutor = ThreadPoolExecutor(1, thread_name_prefix="Pool_%s" % self.name)
 
         self._cache = inmanta.agent.cache.AgentCache(self)
+        # This lock ensures cache entry can not be cleaned up when
+        # the executor is actively working and vice versa
+        self.wip_lock = asyncio.Lock()
```
Comment (sanderr) [inline src/inmanta/agent/in_process_executor.py:73]: I find `wip` a strange name. How about `activity_lock`?

File: src/inmanta/agent/in_process_executor.py [hunk 10]
```diff
@@ -90,6 +112,8 @@ def join(self, thread_pool_finalizer: list[ThreadPoolExecutor]) -> None:
         assert self._stopped
         thread_pool_finalizer.append(self.provider_thread_pool)
         thread_pool_finalizer.append(self.thread_pool)
+        if self.periodic_cache_cleanup_job:
+            await self.periodic_cache_cleanup_job
```
Comment (sanderr) [inline src/inmanta/agent/in_process_executor.py:116]: I believe this will suffer from the same issue as raised by @arnaudsjs [here](https://inmanta.slack.com/archives/CKRF0C8R3/p1721120710485599).

**Review by sanderr — CHANGES_REQUESTED** [review]

File: tests/test_cache.py [hunk 11]
```diff
@@ -39,37 +42,72 @@ class MyResource(Resource):
...
+    )
+
+    cache = myagent_instance._cache
+    value = "test too"
+    cache.cache_value("test", value, timeout=0.1)
+    cache.cache_value("test2", value)
+
+    assert value == cache.find("test")
+    # Cache cleanup job is periodically triggered with a 1s delay
+    await asyncio.sleep(2)
```
[the remark this answers is not in the document]
Comment (Hugo-Inmanta) [reply tests/test_cache.py:75]: Added the `agent_cache_cleanup_tick_rate` config option to set this interval. 
Added a fixture (`set_custom_cache_cleanup_policy`) that sets this interval to 0.1s for this test. I'm monkeypatching the validator in this fixture to accept a float value, I'm not sure this is the cleanest approach.

File: src/inmanta/agent/in_process_executor.py [hunk 12]
```diff
@@ -68,20 +68,42 @@ def __init__(
...
+        self.periodic_cache_cleanup_job = asyncio.create_task(self.cleanup_stale_cache_entries())
+
+    async def cleanup_stale_cache_entries(self) -> None:
+        """
+        Periodically cleans up stale entries in the cache. The clean_stale_entries
+        has to be called on the thread pool because it might call finalizers.
+        """
+
+        while not self._stopped:
+            reschedule_interval: float = 1.0
```
[the remark this answers is not in the document]
Comment (Hugo-Inmanta) [reply src/inmanta/agent/in_process_executor.py]: I'm not sure actually, your other [comment](https://github.com/inmanta/inmanta-core/pull/7846#discussion_r1677892873) about the test being slow with a long expiry interval made me realize it would be nice to add an option to control this interval easily. I used the `is_time` validator for the option (which I had to monkey patch in the test to allow for <1s interval).
Maybe it would be less convoluted to use float everywhere then ?

File: src/inmanta/agent/in_process_executor.py [hunk 13]
```diff
@@ -68,20 +68,51 @@ def __init__(
...
+        """
+        Periodically cleans up stale entries in the cache. The clean_stale_entries
+        has to be called on the thread pool because it might call finalizers.
+        """
+        reschedule_interval: int = self.cache_cleanup_tick_rate
+        while not self._stopped:
+            async with self.activity_lock:
+                if self._stopped:
+                    return
+                await asyncio.get_running_loop().run_in_executor(self.thread_pool, self._cache.clean_stale_entries)
```
Comment (arnaudsjs) [inline src/inmanta/agent/in_process_executor.py:97]: Let's add a try-catch around this to make sure we don't drop out of the while loop if an exception occurs.

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: src/inmanta/agent/in_process_executor.py [hunk 14]
```diff
@@ -68,20 +68,55 @@ def __init__(
...
+        reschedule_interval: int = self.cache_cleanup_tick_rate
+        while not self._stopped:
+            async with self.activity_lock:
+                if self._stopped:
+                    return
+                try:
+                    await asyncio.get_running_loop().run_in_executor(self.thread_pool, self._cache.clean_stale_entries)
+                except Exception:
+                    # Make sure we don't drop out of the while loop if an exception occurs.
+                    pass
```
Comment (arnaudsjs) [inline src/inmanta/agent/in_process_executor.py:101]: Let's also log a message. Otherwise we are not aware about the fact that there is an issue.
```suggestion
                    # Make sure we don't drop out of the while loop if an exception occurs.
                    LOGGER.exception("An unexpected exception occurred while cleaning up the cache of agent %s.", self.name)
```

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: src/inmanta/agent/config.py [hunk 15]
```diff
@@ -211,6 +211,14 @@ def is_executor_mode(value: str | AgentExecutorMode) -> AgentExecutorMode:
     is_time,
 )
 
+agent_cache_cleanup_tick_rate = Option[int](
+    "agent",
+    "cache-cleanup-tick-rate",
+    1,
+    "The rate (in seconds) at which the agent will periodically attempt to remove stale entries from the cache.",
```
Comment (wouterdb) [inline src/inmanta/agent/config.py:218]:
```suggestion
    "The rate (in seconds) at which the agent will periodically attempt to remove stale entries from the cache when idle.",
```

File: tests/test_cache.py [hunk 16]
```diff
@@ -39,37 +44,89 @@ class MyResource(Resource):
...
+        "agent1", "local:", [executor.ResourceInstallSpec("test::Test", 5, blueprint1)]
+    )
+
+    cache = myagent_instance._cache
+    value = "test too"
+    cache.cache_value("test", value, timeout=0.1)
+    cache.cache_value("test2", value)
+
+    assert value == cache.find("test")
+    # Cache cleanup job is periodically triggered with a 1s delay
```
Comment (wouterdb) [inline tests/test_cache.py:93]:
```suggestion
    # Cache cleanup job is periodically triggered with a 0.1s delay
```

**Review by wouterdb — APPROVED** [review]
Nice

File: tests/forking_agent/test_executor.py [hunk 17]
```diff
@@ -224,8 +224,12 @@ async def oldest_gone():
 
     await simplest.request_shutdown()
     await simplest.join()
+
+    async def check_connection_lost() -> bool:
+        return await simplest.call(GetName()) != simplest_blueprint.blueprint_hash()
+
     with pytest.raises(ConnectionLost):
-        await simplest.call(GetName())
+        await retry_limited(check_connection_lost, 1)
```
Comment (Hugo-Inmanta) [inline tests/forking_agent/test_executor.py:232]: Comment for linking purpose
